src/boot/bootctl-install.c

```diff
@@ -569,6 +574,161 @@ static int install_entry_token(void) {
+
+        r = localtime_or_gmtime_usec(epoch != UINT64_MAX ? epoch : now(CLOCK_REALTIME), /*utc=*/ false, &tm);
+        if (r < 0)
+                return log_error_errno(r, "Failed to convert timestamp to calendar time: %m");
+
+        *ret = (EFI_TIME) {
+                .Year = 1900 + tm.tm_year,
+                .Month = tm.tm_mon,
+                .Day = tm.tm_mday,
+                .Hour = tm.tm_hour,
+                .Minute = tm.tm_min,
+                .Second = tm.tm_sec,
+        };
+
+        return 0;
+}
+
+static int install_secure_boot_auto_enroll(const char *esp, X509 *certificate, EVP_PKEY *private_key) {
+#if HAVE_OPENSSL
+        int r;
+
+        _cleanup_free_ uint8_t *dercert = NULL;
+        int dercertsz;
+        dercertsz = i2d_X509(certificate, &dercert);
+        if (dercertsz < 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to convert X.509 certificate to DER: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (arg_private_key_source_type == OPENSSL_KEY_SOURCE_FILE) {
+                r = parse_path_argument(arg_private_key, /*suppress_root=*/ false, &arg_private_key);
+                if (r < 0)
+                        return r;
+        }
+
+        _cleanup_close_ int keys_fd = chase_and_open("loader/keys/auto", esp, CHASE_PREFIX_ROOT|CHASE_PROHIBIT_SYMLINKS, O_DIRECTORY, NULL);
+        if (keys_fd < 0)
+                return log_error_errno(keys_fd, "Failed to chase loader/keys/auto in the ESP: %m");
+
+        _cleanup_free_ EFI_SIGNATURE_LIST *siglist = malloc(sizeof(EFI_SIGNATURE_LIST) + sizeof(EFI_SIGNATURE_DATA) + dercertsz);
+        if (!siglist)
+                return log_oom();
+
+        *siglist = (EFI_SIGNATURE_LIST) {
+                .SignatureType = EFI_CERT_X509_GUID,
+                .SignatureListSize = sizeof(EFI_SIGNATURE_LIST) + sizeof(EFI_SIGNATURE_DATA) + dercertsz,
+                .SignatureHeaderSize = 0,
+                .SignatureSize = sizeof(EFI_SIGNATURE_DATA) + dercertsz,
+        };
+
+        memcpy(siglist->Signatures[0].SignatureData, dercert, dercertsz);
+
+        EFI_TIME timestamp;
+        r = efi_timestamp(&timestamp);
+        if (r < 0)
+                return r;
+
+        uint32_t attrs =
+                EFI_VARIABLE_NON_VOLATILE|
+                EFI_VARIABLE_BOOTSERVICE_ACCESS|
+                EFI_VARIABLE_RUNTIME_ACCESS|
+                EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS;
+
+        FOREACH_STRING(db, "PK", "KEK", "db") {
+                _cleanup_(BIO_freep) BIO *bio = NULL;
+
+                bio = BIO_new(BIO_s_mem());
+                if (!bio)
+                        return log_oom();
+
+                _cleanup_free_ char16_t *db16 = utf8_to_utf16(db, SIZE_MAX);
+                if (!db16)
+                        return log_oom();
+
+                if (BIO_write(bio, db16, char16_strlen(db16) * sizeof(char16_t)) < 0)
```

Comment:
in efi land we have strsize16() for calculating this. maybe we should generalize that and have that in userspace too?